### PR TITLE
Remove support for -swift-version 3 from the expression type checker.

### PR DESCRIPTION
**Pull request by rudkx**
This is the obvious stuff. There will probably be more fallout.

### CODE REVIEW

**Comment by rudkx** [comment]
There are no test changes here, just updates to the code, so this won't pass tests.

Waiting on https://github.com/apple/swift/pull/17689 and any other test updates before merging.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -3689,19 +3688,6 @@ namespace {
...
                                          [&](Expr *sub, Type toInstanceType) {
-        // Warn about NSNumber and NSValue bridging coercions we accepted in
-        // Swift 3 but which can fail at runtime.
-        if (tc.Context.LangOpts.isSwiftVersion3()
-            && tc.typeCheckCheckedCast(cs.getType(sub), toInstanceType,
-                                       CheckedCastContextKind::None,
-                                       dc, SourceLoc(), sub, SourceRange())
-                 == CheckedCastKind::Swift3BridgingDowncast) {
-          tc.diagnose(expr->getLoc(),
-                      diag::missing_forced_downcast_swift3_compat_warning,
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp]: Remove "missing_forced_downcast_swift3_compat_warning"

File: lib/Sema/MiscDiagnostics.cpp [hunk 2]
```diff
@@ -543,33 +535,6 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
...
-      if (isa<TupleExpr>(arg) || isa<TupleShuffleExpr>(arg))
-        return;
-
-      // We care about whether the parameter list of the callee syntactically
-      // has more than one argument.  It has to *syntactically* have a tuple
-      // type as its argument.  A ParenType wrapping a TupleType is a single
-      // parameter.
-      auto params = FT->getParams();
-      if (params.size() > 1) {
-        TC.diagnose(Call->getLoc(), diag::tuple_splat_use, params.size())
```
Comment (slavapestov) [inline lib/Sema/MiscDiagnostics.cpp]: Remove "tuple_splat_use"

File: lib/Sema/MiscDiagnostics.cpp [hunk 3]
```diff
@@ -676,9 +641,7 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
 
       // In Swift 3, this is just a warning.
       TC.diagnose(apply->getLoc(),
-                  TC.Context.isSwiftVersion3()
-                    ? diag::warn_noescape_param_call
```
Comment (slavapestov) [inline lib/Sema/MiscDiagnostics.cpp]: Remove "warn_noescape_param_call"

File: lib/Sema/MiscDiagnostics.cpp [hunk 4]
```diff
@@ -741,36 +704,6 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
 
     // Swift 3 mode produces a warning + Fix-It for the missing ".self"
```
Comment (slavapestov) [inline lib/Sema/MiscDiagnostics.cpp]: Remove this comment

File: lib/Sema/MiscDiagnostics.cpp [hunk 5]
```diff
@@ -797,19 +730,6 @@ static void diagSyntacticUseRestrictions(TypeChecker &TC, const Expr *E,
         }
       }
 
-      if (shouldWarnOnMissingSelf(E)) {
-        auto diag = TC.diagnose(E->getEndLoc(),
-                                diag::warn_value_of_metatype_missing_self,
```
Comment (slavapestov) [inline lib/Sema/MiscDiagnostics.cpp]: Remove "warn_value_of_metatype_missing_self"

**Review by slavapestov — COMMENTED** [review]
Nice work. If you don't mind, can you please remove the dead diagnostics now also?

**Comment by rudkx** [comment]
@slavapestov I think I took care of the things you mentioned.

File: lib/Sema/CSApply.cpp [hunk 6]
```diff
@@ -6545,7 +6501,7 @@ Expr *ExprRewriter::coerceToType(Expr *expr, Type toType,
       // ```
       //
       // See also: https://bugs.swift.org/browse/SR-6796
-      if (cs.getASTContext().isSwiftVersionAtLeast(3) &&
+      if (cs.getASTContext().isSwiftVersionAtLeast(4) &&
```
Comment (slavapestov) [inline lib/Sema/CSApply.cpp]: Isn’t the first part always true? It should suffice to just check if you’re < 5

**Review by slavapestov — COMMENTED** [review]
One more nitpick

File: lib/Sema/CSApply.cpp (hunk 6 above)
Comment (rudkx) [reply lib/Sema/CSApply.cpp]: That was intentional. If there are tests passing -swift-version 3 explicitly they should break if they are relying on this behavior. You might have removed all of those with your PR, though.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@nathawes FYI, I'm removing a few migrator tests here that don't make sense with Swift 3 support removed from the expression type checker.

@benlangmuir I'm not sure what's going on with the code completion test. I'll take a look at it, but in the mean time do you have any idea why removing support for some Swift 3 compatibility hacks would have an impact here?

**Comment by rudkx** [comment]
I've split all the test updates out into https://github.com/apple/swift/pull/17748.

**Comment by rudkx** [comment]
Please test with following PR:
https://github.com/apple/swift/pull/17748

@swift-ci Please smoke test

**Comment by rudkx** [comment]
Please test with following PR:
https://github.com/apple/swift-source-compat-suite/pull/197

@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
Hmmm, neither of my "test with" seems to have actually tested with. I guess testing with another PR from the same repo might not be supported, but I don't know what's up with the source compatibility suite.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
Please test with following pull request:
https://github.com/apple/swift-source-compat-suite/pull/197

@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
Source compatibility failures in 3.x configs:

```
Kickstarter-ReactiveExtensions
ReactiveCocoa
RxDataSources
Starscream
```

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci build toolchain

**Comment by rudkx** [comment]
Linux test failure is unrelated to my changes: `TestFoundation/TestProcess.swift:410: error: TestProcess.test_preStartEndState : XCTAssertEqual failed: ("exit") is not equal to ("uncaughtSignal") - `

**Comment by rudkx** [comment]
@swift-ci Please smoke test Linux platform

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Pull request merged**